lua/autorun/server/sv_pvp_movespeed.lua

```diff
@@ -102,21 +97,45 @@ pvpMoveSpeed.getPlayerWeight = getPlayerWeight
 -- Hook Functions --
 local function onEquip( wep, ply )
     if not isValidPlayer( ply ) then return end
+    if hook.Run( "CFC_PlyMS_CanChangeMoveSpeed", ply ) == false then return end
     local totalWeight = getPlayerWeight( ply ) + getWeaponWeight( wep )
 
     setSpeedFromWeight( ply, totalWeight )
 end
 
 local function onDrop( ply, wep )
     if not isValidPlayer( ply ) then return end
+    if hook.Run( "CFC_PlyMS_CanChangeMoveSpeed", ply ) == false then return end
     local totalWeight = getPlayerWeight( ply ) - getWeaponWeight( wep )
 
     setSpeedFromWeight( ply, totalWeight )
 end
 
+local function onForcedWeigh( ply )
+    if not isValidPlayer( ply ) then return end
+    if hook.Run( "CFC_PlyMS_CanChangeMoveSpeed", ply ) == false then return end
+    local totalWeight = getPlayerWeight( ply )
+
+    setSpeedFromWeight( ply, totalWeight )
+end
+
 -- Hooks --
-hook.Remove( "WeaponEquip", generateCFCHook( "HandleEquipMS" ) )
-hook.Add( "WeaponEquip", generateCFCHook( "HandleEquipMS" ), onEquip )
+hook.Add( "WeaponEquip", "CFC_PlyMS_HandleEquipMS", onEquip )
+
+hook.Add( "PlayerDroppedWeapon", "CFC_PlyMS_HandleDroppedWeaponMS", onDrop )
+
+hook.Add( "CFC_PlyMS_WeighPlayer", "CFC_PlyMS_ForcedWeigh", onForcedWeigh )
+
+-- Powerup Hooks --
+hook.Add( "CFC_Powerups_PowerupRemoved", "CFC_PlyMS_PowerupRemoved", function( ply, powerupId )
+    if powerupId ~= "powerup_speed" then return end
+    if not isValidPlayer( ply ) then return end
+    if not ply:Alive() then return end
+
+    hook.Run( "CFC_PlyMS_WeighPlayer", ply )
+end )
 
-hook.Remove( "PlayerDroppedWeapon", generateCFCHook( "HandleDroppedWeaponMS" ) )
-hook.Add( "PlayerDroppedWeapon", generateCFCHook( "HandleDroppedWeaponMS" ), onDrop )
+hook.Add( "CFC_PlyMS_CanChangeMoveSpeed", "CFC_PlyMS_AccountForSpeedPowerup", function( ply )
```

Comment:
the  whole point of using a hook for this check is so we can define the checks in another addon.